frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToRemote.java

```diff
@@ -0,0 +1,221 @@
+    val pushApprovalDialogBuilder = MessageDialogBuilder.yesNoCancel(
+        getString("action.GitMachete.TraverseAction.dialog.push-verification.title"),
+        getString("action.GitMachete.TraverseAction.dialog.push-verification.untracked.text.HTML")
+            .fmt(gitManagedBranch.getName()))
+        .cancelText(getString("action.GitMachete.TraverseAction.dialog.cancel-traverse"));
+
+    switch (pushApprovalDialogBuilder.show(project)) {
+      case MessageConstants.YES :
+        new GitPushDialog(project, List.of(gitRepository), GitPushSource.create(localBranch),
+            /* isForcePushRequired */ false)
+                .show();
+        break;
+
+      case MessageConstants.NO :
+        break;
+
+      default :
+        return false;
+    }
+    return true;
+  }
+
+  @UIEffect
+  private boolean handleAheadOfRemote(IManagedBranchSnapshot gitMacheteBranch, GitLocalBranch localBranch) {
+    val remoteTrackingBranch = gitMacheteBranch.getRemoteTrackingBranch();
+    assert remoteTrackingBranch != null : "remoteTrackingBranch is null";
+    val pushApprovalDialogBuilder = MessageDialogBuilder.yesNoCancel(
+        getString("action.GitMachete.TraverseAction.dialog.push-verification.title"),
+        getString("action.GitMachete.TraverseAction.dialog.push-verification.ahead.text.HTML")
+            .fmt(gitMacheteBranch.getName(), remoteTrackingBranch.getName()))
+        .cancelText(getString("action.GitMachete.TraverseAction.dialog.cancel-traverse"));
+
+    switch (pushApprovalDialogBuilder.show(project)) {
+      case MessageConstants.YES :
+        new GitPushDialog(project, List.of(gitRepository), GitPushSource.create(localBranch),
+            /* isForcePushRequired */ false)
+                .show();
+        break;
+
+      case MessageConstants.NO :
+        break;
+
+      default :
+        return false;
+    }
+    return true;
+  }
+
+  @UIEffect
+  private boolean handleDivergedFromRemote(IManagedBranchSnapshot gitMacheteBranch,
+      SyncToRemoteStatus syncToRemoteStatus,
+      GitLocalBranch localBranch) {
+    val remoteTrackingBranch = gitMacheteBranch.getRemoteTrackingBranch();
+    assert remoteTrackingBranch != null : "remoteTrackingBranch is null";
+    val selectedAction = new DivergedFromRemoteDialog(project, remoteTrackingBranch, gitMacheteBranch,
+        syncToRemoteStatus).showAndGetThePreferredAction();
+    if (selectedAction == null) {
+      return false;
+    }
+    switch (selectedAction) {
+      case FORCE_PUSH :
+        new GitPushDialog(project, List.of(gitRepository), GitPushSource.create(localBranch),
+            /* isForcePushRequired */ true).show();
+        break;
+
+      case RESET_ON_REMOTE :
+        new ResetCurrentToRemoteBackgroundable(
+            getString("action.GitMachete.BaseResetToRemoteAction.task-title"),
+            /* canBeCancelled */ true, gitMacheteBranch.getName(), remoteTrackingBranch.getName(),
+            gitRepository);
```

Comment:
Maybe here, it might be necessary to invoke or `queue`.

Also, the override of `onSuccess` might be necessary to make sure the next branch, which might be a child of the current one, does not sync to parent before the reset of parent on its remote gets finished.

Reply:
good point about missing `queue`! thanks!

Reply:
hmmm 🤔wouldn’t the run method itself be run concurrently to the existing thread?
or I might be missing something 😕

Reply:
yes, you are right, but thanks to backgroundables we have `onSuccess` to get such a job done! we are cool 🚀

Reply:
in other words, it can be concurrent - we only need a guaranty that `run` finishes what need to be finished and then runs `onSuccess`. that is what happens

Reply:
but there is the other "issue" with reset... I will start this discussion in the traverse feature issue

Reply:
oops, the `onSuccess` does not seems to be overrided in `ResetCurrentToRemoteBackgroundable` or I might be missing sth🧐